### ISSUE TITLE
Batched propagation of linked facts only works for the first set of facts

### ISSUE THREAD

**Issue by snikolayev** [NRules/NRules#338]
When using batched propagation of linked facts (see #241), linked facts queued for propagation are supposed to be grouped into sets, based on the action (insert, update, retract).
E.g. 
`update, update, update, insert, insert, update, retract, retract, retract`
is supposed to become:
`update[3], insert[2], update, retract[3]`
but instead it becomes:
`update[3], insert, insert, update, retract, retract, retract`
In other words, the first consecutive group of updates are batched together, while others remain ungrouped, leading to a potential performance issue.
